source/server/config/http/ext_authz.cc

```diff
@@ -0,0 +1,68 @@
+#include "server/config/http/ext_authz.h"
+
+#include <chrono>
+#include <string>
+
+#include "envoy/config/filter/http/ext_authz/v2/ext_authz.pb.validate.h"
+#include "envoy/registry/registry.h"
+
+#include "common/ext_authz/ext_authz_impl.h"
+#include "common/http/filter/ext_authz.h"
+#include "common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Server {
+namespace Configuration {
+
+HttpFilterFactoryCb ExtAuthzFilterConfig::createFilter(
+    const envoy::config::filter::http::ext_authz::v2::ExtAuthz& proto_config, const std::string&,
+    FactoryContext& context) {
+
+  ASSERT((proto_config.grpc_service().has_envoy_grpc() &&
+          !proto_config.grpc_service().envoy_grpc().cluster_name().empty()) ||
+         (proto_config.grpc_service().has_google_grpc() &&
+          !proto_config.grpc_service().google_grpc().target_uri().empty()));
+
+  Http::ExtAuthz::FilterConfigSharedPtr filter_config(
+      new Http::ExtAuthz::FilterConfig(proto_config, context.localInfo(), context.scope(),
+                                       context.runtime(), context.clusterManager()));
+  const uint32_t timeout_ms = PROTOBUF_GET_MS_OR_DEFAULT(proto_config.grpc_service(), timeout, 200);
+
+  return [ grpc_service = proto_config.grpc_service(), &context, filter_config,
+           timeout_ms ](Http::FilterChainFactoryCallbacks & callbacks)
+      ->void {
+
+    auto async_client_factory =
```

Comment:
Can the factory be made outside the lambda and re-used for all calls?

Reply:
I tried to move it out of the lambda. But it's a `unique_ptr` and is only used within the lamda and not passed into the function called within it.
So the compiler was complaining about it's deletion. Not sure how to handle this.

Reply:
You can move it outside the lambda, then move the unique_ptr into the lambda.  See https://stackoverflow.com/questions/8236521/how-to-capture-a-unique-ptr-into-a-lambda-expression

Reply:
@ggreenway Thanks for the reference. 
I had tried with `move` but i think the problem is that the `unique_ptr` is not absorbed/moved within the lambda. Or to put it in another way the pointer is used __only__ inside the lamda and not passed on to an api called inside the lambda.
For example the diff set
```
diff --git a/source/server/config/http/ext_authz.cc b/source/server/config/http/ext_authz.cc
index eb63ab3..f3ac7da 100644
--- a/source/server/config/http/ext_authz.cc
+++ b/source/server/config/http/ext_authz.cc
@@ -28,12 +28,12 @@ HttpFilterFactoryCb ExtAuthzFilterConfig::createFilter(
       context.clusterManager());
   const uint32_t timeout_ms = PROTOBUF_GET_MS_OR_DEFAULT(proto_config.grpc_service(), timeout, 200);
 
-  return [ grpc_service = proto_config.grpc_service(), &context, filter_config,
+  auto async_client_factory =
+        context.clusterManager().grpcAsyncClientManager().factoryForGrpcService(proto_config.grpc_service(),
+                                                                                context.scope());
+  return [ async_client_factory = std::move(async_client_factory) , &context, filter_config,
            timeout_ms ](Http::FilterChainFactoryCallbacks & callbacks)
       ->void {
-    auto async_client_factory =
-        context.clusterManager().grpcAsyncClientManager().factoryForGrpcService(grpc_service,
-                                                                                context.scope());
     auto client = std::make_unique<Envoy::ExtAuthz::GrpcClientImpl>(
         async_client_factory->create(), std::chrono::milliseconds(timeout_ms));
     callbacks.addStreamDecoderFilter(Http::StreamDecoderFilterSharedPtr{
```
throws the following compile error:
```
                                  ^
source/server/config/http/ext_authz.cc:35:23: note: 'Envoy::Server::Configuration::ExtAuthzFilterConfig::createFilter(const envoy::config::filter::http::ext_authz::v2::ExtAuthz&, const string&, Envoy::Server::Configuration::FactoryContext&)::<lambda(Envoy::Http::FilterChainFactoryCallbacks&)>::<lambda>(const Envoy::Server::Configuration::ExtAuthzFilterConfig::createFilter(const envoy::config::filter::http::ext_authz::v2::ExtAuthz&, const string&, Envoy::Server::Configuration::FactoryContext&)::<lambda(Envoy::Http::FilterChainFactoryCallbacks&)>&)' is implicitly deleted because the default definition would be ill-formed:
            timeout_ms ](Http::FilterChainFactoryCallbacks & callbacks)
                       ^
source/server/config/http/ext_authz.cc:35:23: error: use of deleted function 'std::unique_ptr<_Tp, _Dp>::unique_ptr(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = Envoy::Grpc::AsyncClientFactory; _Dp = std::default_delete<Envoy::Grpc::AsyncClientFactory>]'
In file included from /usr/include/c++/5/memory:81:0,
                 from external/com_google_protobuf/src/google/protobuf/descriptor.h:57,
                 from external/com_google_protobuf/src/google/protobuf/message.h:122,
                 from bazel-out/k8-fastbuild/genfiles/external/envoy_api/envoy/config/filter/http/ext_authz/v2/ext_authz.pb.h:28,
                 from bazel-out/k8-fastbuild/bin/source/server/config/http/_virtual_includes/ext_authz_lib/server/config/http/ext_authz.h:5,
                 from source/server/config/http/ext_authz.cc:1:
/usr/include/c++/5/bits/unique_ptr.h:356:7: note: declared here
       unique_ptr(const unique_ptr&) = delete;
```

Reply:
Oh, I see. The problem is that if the std::unique_ptr is moved into the lambda, the lambda itself becomes non-copyable; it is only movable.  Logically we don't need to copy the lambda, we only need 1 copy of it, but std::function doesn't really work this way.  I think the only way around this would be to make factoryForGrpcService() return a shared_ptr instead of a unique_ptr, or switch from std::function to something that uses move instead of copy semantics. @htuch any thoughts on this?

I think for this PR, let's just leave this how it is.

Reply:
@ggreenway Thanks for the explanation! At least the part about the impact of moving a  `std::unique_ptr` in to the lambda makes sense.